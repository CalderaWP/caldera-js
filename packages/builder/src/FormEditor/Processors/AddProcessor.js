import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Row, fieldAreaFactory} from "@calderajs/components";

const AddProcessorButtonText = ({processorTypes,value}) => {
	const findProcessorLabel = type => {
		const processor = processorTypes.find( processorType => value === processorType.type );
		if(  processor && processor.hasOwnProperty('typeLabel') ){
			return processor.typeLabel;
		}
		return type; 

	}
	if( ! value ){
		return <Fragment>Choose Processor Type</Fragment>
	}

	const label = findProcessorLabel(value);
	return <Fragment>{ `Add ${label} Processor`}</Fragment>
};

export const AddProcessor = (
	{
		setNewProcessorType,
		processorTypes,
		onCreate,
		children,
		value
	}
) => {
	const options = [];
	console.log(processorTypes)
	processorTypes.map(processorType => {
		options.push({
			value: processorType.type,
			label: processorType.hasOwnProperty( 'typeLabel') ? processorType.typeLabel : processorType.type
		});
	});

	const processorTypesField = {
		fieldType: 'select',
		label: 'Processor Type',
		fieldId: 'newProcessorType',
		required: true,
		options,
		onChange: setNewProcessorType,
		value,
	};

	
	

	return (
		<Fragment>
			{fieldAreaFactory(processorTypesField, setNewProcessorType)}
			<button 
				onClick={onCreate}
				disabled={!value}
			>
				<AddProcessorButtonText value={value} processorTypes={processorTypes}/>
			</button>
		</Fragment>
	)
};

AddProcessor.propTypes = {
	processorTypes: PropTypes.array,
	onCreate: PropTypes.func,
	setNewProcessorType: PropTypes.func,
	value: PropTypes.string
};

AddProcessor.defaultProps = {
	processorTypes: [],
	value: ''
};
