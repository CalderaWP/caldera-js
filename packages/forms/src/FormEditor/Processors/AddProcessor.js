import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Row, fieldAreaFactory} from "@calderawp/components";

export const AddProcessor = ({
								 setNewProcessorType,
								 processorTypes,
								 onCreate,
								 children,
								 value
							 }) => {
	const options = [
		{
			value: null,
			label: '--'
		}
	];

	processorTypes.map(processorType => {
		options.push({
			value: processorType.type,
			label: processorType.type
		});
	});


	const processorTypesField = {
		fieldType: 'select',
		label: 'Processor Type',
		fieldId: 'newProcessorType',
		required: true,
		options,
		onChange: setNewProcessorType,
		value,
	};


	return (
		<Fragment>
			{fieldAreaFactory(processorTypesField, setNewProcessorType)}
			<button onClick={onCreate}
					disabled={!value}
			>
				{children}
			</button>
		</Fragment>
	)
};

AddProcessor.propTypes = {
	processorTypes: PropTypes.array,
	onCreate: PropTypes.func,
	setNewProcessorType: PropTypes.func,
	value: PropTypes.string
};

AddProcessor.defaultProps = {
	processorTypes: [],
	value: ''
};
