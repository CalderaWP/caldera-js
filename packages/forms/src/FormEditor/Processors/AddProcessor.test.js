import React, {Component} from 'react';
import {mount} from 'enzyme';
import renderer from 'react-test-renderer';
import {AddProcessor} from './AddProcessor'

describe('AddProcessor', () => {
	let setNewProcessorType = jest.fn();
	let onCreate = jest.fn();
	beforeEach(() => {
		setNewProcessorType = jest.fn();
		onCreate = jest.fn();
	});
	const processorTypes = [
		{
			type: 'apiRequest'
		},
		{
			type: 'redirect'
		}
	];

	it('Shows the choices', () => {
		const component = renderer.create(<AddProcessor
			setNewProcessorType={setNewProcessorType}
			processorTypes={processorTypes}
			value={''}
			onCreate={onCreate}
		/>);
		expect(component.toJSON()).toMatchSnapshot()
	});

	it('Activates the button', () => {
		const component = mount(<AddProcessor
			setNewProcessorType={setNewProcessorType}
			processorTypes={processorTypes}
			value={'redirect'}
			onCreate={onCreate}
		/>);
		expect(component.find( 'button').prop( 'disabled')).toBe(false)

	})

	it('Has the processor options', () => {
		const component = mount(<AddProcessor
			setNewProcessorType={setNewProcessorType}
			processorTypes={processorTypes}
			value={'redirect'}
			onCreate={onCreate}
		/>);
		expect(component.find( 'select').find( 'option').length ).toBe(processorTypes.length + 1 );

	})
});
