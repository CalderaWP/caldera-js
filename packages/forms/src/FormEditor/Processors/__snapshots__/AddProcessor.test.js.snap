// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProcessor Shows the choices 1`] = `
Array [
  <div
    className="caldera-field-wrapper caldera-field-wrapper-text"
>
    <div
        className="components-base-control caldera-field caldera-field-select"
    >
        <div
            className="components-base-control__field"
        >
            <label
                className="components-base-control__label"
                htmlFor="inspector-select-control-0"
            >
                Processor Type
            </label>
            <select
                aria-describedby={undefined}
                className="components-select-control__input"
                id="newProcessorType"
                multiple={false}
                onChange={[Function]}
                value=""
            >
                <option
                    value={null}
                >
                    --
                </option>
                <option
                    value="apiRequest"
                >
                    apiRequest
                </option>
                <option
                    value="redirect"
                >
                    redirect
                </option>
            </select>
        </div>
    </div>
</div>,
  <button
    disabled={true}
    onClick={[Function]}
/>,
]
`;
